### PR TITLE
SNOW-2069271:Support mysql ingestion

### PR DESCRIPTION
**Pull request by sfc-gh-yuwang**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-2069271

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://github.com/snowflakedb/snowpark-python/blob/main/CONTRIBUTING.md#thread-safe-development)
   - [x] If adding any arguments to public Snowpark APIs or creating new public Snowpark APIs, I acknowledge that I have ensured my changes include AST support. Follow the link for more information: [AST Support Guidelines](https://github.com/snowflakedb/snowpark-python/blob/main/CONTRIBUTING.md#ast-abstract-syntax-tree-support-in-snowpark)

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/data_source/dbms_dialects/mysql_dialect.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+
+class MysqlDialect(BaseDialect):
+    def generate_select_query(
+        self,
+        table_or_query: str,
+        schema: StructType,
+        raw_schema: List[tuple],
+    ) -> str:
+        is_query = table_or_query.lower().startswith("select")
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/dbms_dialects/mysql_dialect.py:17]: let's introduce a flag to distinct table from query

File: src/snowflake/snowpark/_internal/data_source/dbms_dialects/mysql_dialect.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+                else:
+                    cols.append(f"""CAST(`{raw_field[0]}` AS CHAR) AS {raw_field[0]}""")
+            else:
+                if is_query:
+                    cols.append(f"""A.`{raw_field[0]}` AS {raw_field[0]}""")
+                else:
+                    cols.append(f"`{raw_field[0]}`")
+
+        if is_query:
+            return f"""SELECT {" , ".join(cols)} FROM ({table_or_query}) A"""
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/dbms_dialects/mysql_dialect.py:34]: A might be conflicting other ids in the col/table, let's use a random name "SNOWPARK_***"

File: src/snowflake/snowpark/_internal/data_source/drivers/pymsql_driver.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+            for i, col in enumerate(row):
+                raw_data_types_set[i].add(type(col))
+        # NoneType is not exposed in python <3.10 have to define it ourselves
+        NoneType = type(None)
+        types = [
+            type_set.pop()
+            if len(type_set) == 1 and next(iter(type_set)) != NoneType
+            else str
+            for type_set in raw_data_types_set
+        ]
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/drivers/pymsql_driver.py:217]: we should document this behavior as limitation

File: src/snowflake/snowpark/_internal/data_source/drivers/pymsql_driver.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    def prepare_connection(
+        self,
+        conn: "Connection",
+        query_timeout: int = 0,
+    ) -> "Connection":
+        conn.read_timeout = query_timeout if query_timeout != 0 else None
+        return conn
+
+    @staticmethod
+    def infer_type_from_data(data: List[tuple], number_of_columns: int) -> List[Type]:
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/drivers/pymsql_driver.py:204]: can we make number_of_columns configurable, we can define a module variable

File: src/snowflake/snowpark/_internal/data_source/drivers/pymsql_driver.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+                internal_size,
+                precision,
+                scale,
+                null_ok,
+            ) = col
+            snow_type = BASE_PYMYSQL_TYPE_TO_SNOW_TYPE.get(type_code, None)
+            if snow_type is None:
+                raise NotImplementedError(f"mysql type not supported: {type_code}")
+            if type_code in (PymysqlTypeCode.DECIMAL, PymysqlTypeCode.NEWDECIMAL):
+                precision -= 2
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/drivers/pymsql_driver.py:155]: 1. let's add some comment on why we did -2 here
2. let's call out the limitation as connector that we don't support precision beyond 38

File: src/snowflake/snowpark/_internal/data_source/utils.py [hunk 6]
```diff
@@ -74,6 +82,7 @@ class DRIVER_TYPE(str, Enum):
         "msodbcsql",
         "snowflake-snowpark-python",
     ],
+    DBMS_TYPE.MYSQL_DB: ["pymysql", "snowflake-snowpark-python"],
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/utils.py:85]: let's add version cap, pymysql<=a,>=b
```suggestion
    DBMS_TYPE.MYSQL_DB: ["pymysql", "snowflake-snowpark-python"],
```

File: tests/integ/datasource/test_mysql.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+def test_type_conversion():
+    invalid_type = MysqlType("ID", "UNKNOWN", None, None, None, None, False)
+    with pytest.raises(NotImplementedError, match="mysql type not supported"):
+        PymysqlDriver(create_connection_mysql, DBMS_TYPE.MYSQL_DB).to_snow_type(
+            [invalid_type]
+        )
+
+
+def test_oracledb_driver_coverage(caplog):
```
Comment (sfc-gh-aling) [inline tests/integ/datasource/test_mysql.py:96]: nit: mysql

File: tests/integ/datasource/test_mysql.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+from tests.resources.test_data_source_dir.test_mysql_data import (
+    mysql_real_data,
+    MysqlType,
+)
+from tests.utils import RUNNING_ON_JENKINS, Utils
+from tests.parameters import MYSQL_CONNECTION_PARAMETERS
+
+DEPENDENCIES_PACKAGE_UNAVAILABLE = True
+try:
+    import databricks  # noqa: F401
```
Comment (sfc-gh-aling) [inline tests/integ/datasource/test_mysql.py:22]:
```suggestion
    import pymysql  # noqa: F401
```

File: tests/integ/datasource/test_mysql.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    import pandas  # noqa: F401
+
+    DEPENDENCIES_PACKAGE_UNAVAILABLE = False
+except ImportError:
+    pass
+
+
+pytestmark = [
+    pytest.mark.skipif(DEPENDENCIES_PACKAGE_UNAVAILABLE, reason="Missing 'databricks'"),
```
Comment (sfc-gh-aling) [inline tests/integ/datasource/test_mysql.py:31]:
```suggestion
    pytest.mark.skipif(DEPENDENCIES_PACKAGE_UNAVAILABLE, reason="Missing 'pymysql'"),
```

File: tests/integ/datasource/test_mysql.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        RUNNING_ON_JENKINS, reason="cannot access external datasource from jenkins"
+    ),
+    pytest.mark.skipif(
+        "config.getoption('local_testing_mode', default=False)",
+        reason="feature not available in local testing",
+    ),
+]
+
+TEST_TABLE_NAME = "ALL_TYPES_TABLE"
+MYSQL_TEST_EXTERNAL_ACCESS_INTEGRATION = "mysql_integration"
```
Comment (sfc-gh-aling) [inline tests/integ/datasource/test_mysql.py:42]: nit: I think we should align the naming pattern like other tests, test_..._integration

File: tests/integ/datasource/test_mysql.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+
+def create_connection_mysql():
+    import pymysql
+
+    conn = pymysql.connect(
+        user=MYSQL_CONNECTION_PARAMETERS["username"],
+        password=MYSQL_CONNECTION_PARAMETERS["password"],
+        host=MYSQL_CONNECTION_PARAMETERS["host"],
+        database=MYSQL_CONNECTION_PARAMETERS["database"],
```
Comment (sfc-gh-aling) [inline tests/integ/datasource/test_mysql.py:52]: nit: you can do `pymysql.connect(**MYSQL_CONNECTION_PARAMETERS)`

File: tests/integ/datasource/test_mysql.py [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+@pytest.mark.parametrize(
+    "create_connection, table_name, expected_result",
+    [
+        (
+            create_connection_mysql,
+            TEST_TABLE_NAME,
+            mysql_real_data,
+        ),
+    ],
+)
```
Comment (sfc-gh-aling) [inline tests/integ/datasource/test_mysql.py:71]: let's add test for query as well

File: tests/integ/datasource/test_mysql.py [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+            [int, float, str, str],
+        ),
+    ],
+)
+def test_infer_type_from_data(data, number_of_columns, expected_result):
+    result = PymysqlDriver.infer_type_from_data(data, number_of_columns)
+    assert result == expected_result
+
+
+def test_udtf_ingestion_oracledb(session):
```
Comment (sfc-gh-aling) [inline tests/integ/datasource/test_mysql.py:151]:
```suggestion
def test_udtf_ingestion_pymysql(session):
```

File: tests/integ/datasource/test_mysql.py (hunk 11 above)
Comment (sfc-gh-yuwang) [reply tests/integ/datasource/test_mysql.py:52]: cannot do it for now as the key value in connection params is a bit different from the arg name in pymysql.connect, will log a jira to fix it

File: src/snowflake/snowpark/_internal/data_source/dbms_dialects/base_dialect.py [hunk 14]
```diff
@@ -7,6 +7,9 @@
 
 
 class BaseDialect:
+    def __init__(self, is_query: bool) -> None:
+        self.is_query = is_query
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/dbms_dialects/base_dialect.py:11]: from OO design perspective, it feels a bit weird to me that the Dialect instance has a is_query mem var.

can we move this flag to the static methods?

File: src/snowflake/snowpark/_internal/data_source/drivers/pymsql_driver.py [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+            fields.append(StructField(name, data_type, null_ok))
+        return StructType(fields)
+
+    def udtf_class_builder(self, fetch_size: int = 1000) -> type:
+        create_connection = self.create_connection
+
+        class UDTFIngestion:
+            def process(self, query: str):
+                def hexify_binary(row):
+                    return tuple(v.hex() if isinstance(v, bytes) else v for v in row)
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/drivers/pymsql_driver.py:199]: I think we can push the hex op to mysql: https://dev.mysql.com/doc/refman/8.4/en/string-functions.html#function_hex in mysql dialect

this function should work for binary col

File: src/snowflake/snowpark/_internal/data_source/utils.py [hunk 16]
```diff
@@ -74,6 +82,7 @@ class DRIVER_TYPE(str, Enum):
         "msodbcsql",
         "snowflake-snowpark-python",
     ],
+    DBMS_TYPE.MYSQL_DB: ["pymysql>=1.0.0,<=2.0.0", "snowflake-snowpark-python"],
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/utils.py:85]: major version bump can introduce breaking changes, it should be less than 2
```suggestion
    DBMS_TYPE.MYSQL_DB: ["pymysql>=1.0.0,<2.0.0", "snowflake-snowpark-python"],
```

File: tests/resources/test_data_source_dir/test_mysql_data.py [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+        "type_code",
+        "display_size",
+        "internal_size",
+        "precision",
+        "scale",
+        "null_ok",
+    ],
+)
+
+mysql_real_data = [
```
Comment (sfc-gh-aling) [inline tests/resources/test_data_source_dir/test_mysql_data.py:36]: let's add one more test row with all None/Null data to check the behavior

**Comment by sfc-gh-aling** [comment]
the pipeline failed with
```bash
FAILED tests/integ/datasource/test_mysql.py::test_type_conversion - TypeError: PymysqlDriver.__init__() missing 1 required positional argument:...
FAILED tests/integ/datasource/test_mysql.py::test_pymysql_driver_coverage - TypeError: PymysqlDriver.__init__() missing 1 required positional argument:...
FAILED tests/integ/datasource/test_oracledb.py::test_type_conversion - TypeError: OracledbDriver.__init__() missing 1 required positional argument...
FAILED tests/integ/datasource/test_oracledb.py::test_oracledb_driver_coverage - TypeError: OracledbDriver.__init__() missing 1 required positional argument...
FAILED tests/integ/datasource/test_mysql.py::test_udtf_ingestion_mysql - AssertionError: Mismatch on row 0 at column 29. Expected {
```

File: CHANGELOG.md [hunk 18]
```diff
@@ -21,6 +21,9 @@
 
 ### Snowpark Python API Updates
 
+#### New Features
```
Comment (sfc-gh-aling) [inline CHANGELOG.md:24]: nit: 1.33.0 (this is what you caught in my PR :P)

File: src/snowflake/snowpark/_internal/data_source/drivers/databricks_driver.py [hunk 19]
```diff
@@ -28,25 +32,22 @@
...
-            ) from exc
-        finally:
-            cursor.close()
-            conn.close()
+    def __init__(
+        self,
+        create_connection: Callable[[], "Connection"],
+        dbms_type: Enum,
+    ) -> None:
+        super().__init__(create_connection, dbms_type)
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/drivers/databricks_driver.py:40]: I think we can remove this `__init__` since it's just calling super()?

File: src/snowflake/snowpark/_internal/data_source/drivers/oracledb_driver.py [hunk 20]
```diff
@@ -27,7 +27,9 @@
 
 class OracledbDriver(BaseDriver):
     def __init__(
-        self, create_connection: Callable[[], "Connection"], dbms_type: Enum
+        self,
+        create_connection: Callable[[], "Connection"],
+        dbms_type: Enum,
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/drivers/oracledb_driver.py:32]: same, I think this can be `__init__` definition can be removed

File: src/snowflake/snowpark/_internal/data_source/drivers/pyodbc_driver.py [hunk 21]
```diff
@@ -41,7 +41,9 @@
 
 class PyodbcDriver(BaseDriver):
     def __init__(
-        self, create_connection: Callable[[], "Connection"], dbms_type: Enum
+        self,
+        create_connection: Callable[[], "Connection"],
+        dbms_type: Enum,
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/drivers/pyodbc_driver.py:46]: nit: same, this `__init__` can be removed

File: src/snowflake/snowpark/_internal/data_source/utils.py [hunk 22]
```diff
@@ -14,6 +14,9 @@
     SqlServerDialect,
     DatabricksDialect,
 )
+from snowflake.snowpark._internal.data_source.dbms_dialects.mysql_dialect import (
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/utils.py:17]: should we add `mysql_dialect to dbms_dialects.__init__`?

File: src/snowflake/snowpark/_internal/data_source/utils.py [hunk 23]
```diff
@@ -22,6 +25,7 @@
 )
 import snowflake
 from snowflake.snowpark._internal.data_source import DataSourceReader
+from snowflake.snowpark._internal.data_source.drivers.pymsql_driver import PymysqlDriver
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/utils.py:28]: add `PymysqlDriver` to `drivers.__init__.py` like other drivers?

**Review by sfc-gh-aling — APPROVED** [review]
I left some minor comments, overall the PR LGTM
